app/pages/Recovery/util.ts

```diff
@@ -0,0 +1,304 @@
+import { getCurrentYearMonth } from '~/utils/timeHelpers';
+import { insertIdentity } from '~/database/IdentityDao';
+import { maxCredentialsOnAccount } from '~/constants/recoveryConstants.json';
+import { createAccount } from '~/utils/accountHelpers';
+import { getNextCredentialNumber } from '~/database/CredentialDao';
+import { importAddressBookEntry } from '~/features/AddressBookSlice';
+
+export enum Status {
+    initial = 'Waiting...',
+    waitingForInput = 'Waiting for Ledger input.',
+    searching = 'Looking for accounts...',
+}
+
+export function getRecoveredIdentityName(identityNumber: number) {
+    return `Recovered: Index ${identityNumber}`;
+}
+
+/**
+ * Creates an placeholder identity,
+ * @param walletId the wallet on which the identity is created.
+ * @param identityNumber the identity's number on the wallet.
+ * @returns the id of the created identity, or the id of the already existing identity
+ */
+export async function createRecoveredIdentity(
+    walletId: number,
+    identityNumber: number
+): Promise<number> {
+    const createdAt = getCurrentYearMonth();
+    const validTo = getCurrentYearMonth();
+    const identityObject = {
+        v: 0,
+        value: {
+            attributeList: {
+                chosenAttributes: {},
+                createdAt,
+                validTo,
+            },
+        },
+    };
+
+    const identity = {
+        name: getRecoveredIdentityName(identityNumber),
+        identityNumber,
+        identityObject: JSON.stringify(identityObject),
+        status: IdentityStatus.Recovered,
+        detail: '',
+        codeUri: '',
+        identityProvider: '{}',
+        randomness: '',
+        walletId,
+    };
+
+    return (await insertIdentity(identity))[0];
+}
+
+interface CredentialIndexAndPolicy {
+    credentialIndex?: number;
+    policy: Policy;
+}
+
+/**
+ * Given a credId, and accountInfo, extract the credential corresponding to the credId.
+ * N.B. If the credId is not in the accountInfo, we assume that it has been removed.
+ * @returns a CredentialIndexAndPolicy object. CredentialIndex is undefined if the credential is not in the accountInfo.
+ */
+function getCredentialOnChain(
+    credId: string,
+    accountInfo: AccountInfo
+): CredentialIndexAndPolicy {
+    const credentialOnChain = Object.entries(
+        accountInfo.accountCredentials
+    ).find(
+        ([, cred]) =>
+            (cred.value.contents.credId || cred.value.contents.regId) === credId
+    );
+    if (!credentialOnChain) {
+        return {
+            credentialIndex: undefined,
+            policy: {
+                validTo: getCurrentYearMonth(),
+                createdAt: getCurrentYearMonth(),
+                revealedAttributes: {},
+            },
+        };
+    }
+
+    return {
+        credentialIndex: parseInt(credentialOnChain[0], 10),
+        policy: credentialOnChain[1].value.contents.policy,
+    };
+}
+
+/**
+ * Attempts to recover the credential with the given credId.
```

Comment:
It would be nice if the documentation explained a bit more, i.e. the talk we had about why we still insert removed credentials.